### ISSUE TITLE
Interpreting 'terms at least as permissive' in the Parity Public license

### ISSUE THREAD

**Issue by makkus** [licensezero/parity-public-license#9]
Related to this: https://github.com/licensezero/licensezero-questions/issues/13
I guess I'm not quite sure how I should interpret "terms as least as permissive" in certain cases. 

Say you have a build or project management tool that is Parity licensed, and which touches files of a open-source project (GPL, Apache, etc.). The first question would be whether mere managing of such a project would fall under 'run or combine', but I guess not. I reckon probably build outcomes would though, right? For example if it builds a Docker image that contains the actual project (not the Parity-licensed build tool itself), that Docker image would be covered by the license?

If that's the case, would it be allowed for that Docker image to be licensed under the original project license (GPL, Apache, ...)?

I'm not 100% sure those questions make a whole lot of sense, if not, sorry. Licensing does my head in, I'm not sure how you do it :-) I have to say though, the Parity and Prosperity license texts are quite a bit nicer to read than any of the others I've tried...

**Comment by kemitchell** [comment]
Thank you for opening an issue here!

Color me impressed. Your comments draw straight lines right to the two most vague phrases in Parity:

1. "at least as permissive", in the instructions for how to release source code
2. "analyze, change, or otherwise make using this software", in rule 3

I'll try to keep my comments short. But I also want to give you the whole picture.

In the end, we have to make trade-offs between specificity, approachability, resilience as technology changes, and written length. The two phrases above trade some specificity for the other three.

# At Least as Permissive

Parity was written as a "fork" or rewrite of a prior license called [L0-R](https://github.com/licensezero/licensezero-reciprocal-public-license). The earlier license also had instructions for release of source code, and required licensing on either the same terms, or the terms of a license approved by the Open Source Initiative.

Pointing to OSI's list of approved licenses is a very common way to define "open source" in contracts. But having actually gone through their license review process myself, and doing a lot of reading as a result, my opinion on it really soured. I received _some_ truly excellent feedback through OSI's mailing list. But for reasons that go very far beyond what they will or won't do with the license I proposed, I came to see writing OSI into my license as a risk, presently and longer-term.

"Permissive" has an ordinary meaning---granting broad permission---but also a slightly more specific meaning in open source licensing. Licenses like MIT, BSD, and Apache 2, which require attribution and some recordkeeping, like preserving notices with copies, are often called "permissive" licenses. Permissive contrasts with "copyleft", "share-alike", or "reciprocal" licenses, notably GPL, which have more conditions, especially around sharing and licensing source code. On that scale, Parity is even more share-alike, even less permissive, than GPL or even AGPL.

Before you read on, if you have time, I'd really appreciate you taking a look at Parity's text, and sharing your opinion. How does it read _to you_? That matters. In many ways, more than what I say!

That being said, my intent in drafting Parity---which may differ from what folks understand from the terms, which determines how courts will enforce---was to cover most common open source licenses we see today. I would argue that MIT and the two- or three-clause BSD licenses are more permissive than Parity, because they require only attribution, while Parity also requires releasing source code in some cases. I would argue that GPL and even AGPL are more permissive than Parity, because they require sharing and licensing source code in fewer scenarios.

Many open source licenses have conditions that are in some sense just _different_ from Parity's. For example, GPL and Apache 2.0 have conditions that require summarizing changes made to original or "official" releases. Artistic 2.0 has rules about not fighting with the original project for executable and package names. One of my long-term projects is cataloging and coding literally every condition in every open source license, so that I can try to generalize them. But Parity isn't going to include its own precise, functional definition of what counts as an open source license. Despite differences in kind among some common open source license conditions, I would read licenses like Apache, with administrative notice and other conditions, as "more permissive" than Parity.

# Analyze, Change, Make

An important goal of Parity is to offer an effective share-alike license for developers writing software development tools. While drafting, I've kept a list of example projects, like ESLint, Babel, Closure Compiler, Browserify, and so on, to use as test cases. Rule 3 is for them.

An important _limitation_ on Rule 3, as short as it reads, is "software". If work product can't naturally called "software", or it doesn't have any "source code" form, Parity doesn't require that it be released.

As I thought about how to write Rule 3, I kept returning to a "test case" that bothered me: Sublime Text, or any text editor. We use text editors to make source code. Did I really want to offer a license that would require users of a text editor to release the code they write with the editor? I'm not sure. But in the end, I decided that wasn't for a license drafter to decide. That was for editor programmers to decide.

You've given me another test case of that kind: a project management tool. In that context, "make" starts to read much more broadly, in the sense that a _team_ makes software. But also in that context, I think the specifically way Rule 3 is written starts to matter:

> 3. Release all source code for software that you analyze, change, or otherwise make using this software.

I used "or" intentionally, to imply that "make" means things like analyzing and changing. The kind of _making_ that a particular developer, not a team or collective, does with a particular tool.

I _do_ think it's up to me, as license drafter, to avoid writing "gotcha" licenses. The point of License Zero isn't to surprise people downloading software from the Internet with heinous license terms, or to require releases of source code as a shock and surprise, when they least expect it. The clearest manifestation of that goal is the automatic forgiveness language at the bottom:

> You are excused for unknowingly breaking 1, 2, or 3 if you release source code, or stop doing anything requiring this license, within 30 days of learning you broke the rule.

But I also expressed it in the language of Rule 3. I don't think folks would expect the license for their project management system to require open source release. Parity is _different_ from the kinds of share-alike licenses we've seen before, but it's still a license for _tools_ rather than platforms. Parity works great for platforms, and requires those making changes to platforms to share them. But it doesn't try to help platform programmers set rules about work that gets discussed on their platforms, rather than made in some more direct sense on their platforms.

I wouldn't be surprise to meet a platform developer someday who changes my mind, for a platform that's almost all about developer communication, rather than writing or analyzing code. In the meantime, some platforms might _include_ tools, like text editors or code analyzers, with Rule 3 implications under the current version of Parity.

**Comment by makkus** [comment]
Thanks for the in-depth answer, much appreciated!

>   On that scale, Parity is even more share-alike, even less permissive, than GPL or even AGPL.

Yes, I see that now, the way you put it makes sense.

> Before you read on, if you have time, I'd really appreciate you taking a look at Parity's text, and sharing your opinion. How does it read to you? That matters. In many ways, more than what I say!

The first time I've read the license text, I read it the way you explained it just now. Once I re-read it some doubts crept in, which is why I created this issue. I guess I wasn't quite sure how to compare the permissions of those common OSI licenses with the Parity ones. I probably should have looked at the restrictions instead, and I would have been more certain I understood it right. Point 2 and 3 are clearly more restrictive than other common open-source licenses, even AGPL, esp. because they get triggered easier/earlier.

> Many open source licenses have conditions that are in some sense just different from Parity's.

Yes, now that you point that out, I think that's what caused me to doubt my initial reading.

>  But Parity isn't going to include its own precise, functional definition of what counts as an open source license.

I agree, this is a good way of dealing with those differences. In a way, one could say it is easier, or less hassle, or cheaper (whatever that means in a certain situation) to deal with those restrictions than with Paritys Point 2 and 3.

> I used "or" intentionally, to imply that "make" means things like analyzing and changing. The kind of making that a particular developer, not a team or collective, does with a particular tool.

Again, I think this approach makes sense, and in my mind it closes some loopholes that make AGPL useless in some scenarios.

> I wouldn't be surprise to meet a platform developer someday who changes my mind, for a platform that's almost all about developer communication, rather than writing or analyzing code.

Even if that was the case, I think in that scenario picking a different license then Parity would not be the end of the world, as that would probably be a much more a 'consumer' product than a development tool, with a different set of monetization options (providing a hosted version, open-core, etc...). 

I think Parity is a good tool to have at ones disposal to get better control of how to control for what purposes ones code can be used for free while still be kinda open source, even if it doesn't satisfy a purists' interpretation of 'open source'.

Thanks again, I do have a clearer understanding of the important points now!

**Comment by danielcompton** [comment]
FWIW, the only part of the license I found unclear was this, and which licenses qualified. 

> and license contributions not already licensed to the public _on terms as permissive as this license_ accordingly.

When I first read it, for some reason I thought that you had to release your code as Parity or a license with less permissions. I think it’s because Parity is not a very permissive license, so that sentence was hard to parse.  I’m not sure if there is a different way to say that which makes the intent clearer? Or maybe it’s fine and I just read it wrong.

**Comment by kemitchell** [comment]
@danielcompton your confusion is completely justified. And here comes the part where I sheepishly admit we saw it coming. We've definitely discussed the fact that Parity is "permissive" by absolutely standards, but not by currently prevailing open source standards.

We've just kicked off a project to restructure in #54. I was tempted to do some more clarification about which licenses count there straightaway, but didn't. I think I'll go back and push a commit doing so now.

I'll mention you in the commit.
